src/sql/src/func.rs

```diff
@@ -1067,11 +1088,78 @@ where
     })
 }
 
+/// Adjust the types of arguments when they contain both ints and uints.
+///
+/// Without this intervention, these arguments will almost certainly select the
+/// `double`/`Float64` version of a function, e.g. equality. Because this
+/// integer to float conversion is surprising, we instead cast the types of the
+/// narrowest type that can represent both the widest signed and widest unsigned
+/// integers.
+///
+/// This is necessary in Materialize but not in PG because PG's types offer a
+/// linear progression of numbers capable of expressing more significant digits,
+/// while Materialize supports unsigned types which creates a non-linear
+/// structure of type promotion.
+fn adjust_ints_uints<R>(
+    ecx: &ExprContext,
+    impls: &[&FuncImpl<R>],
+    args: &mut [CoercibleScalarExpr],
+    types: &mut Vec<Option<ScalarType>>,
+) where
+    R: fmt::Debug,
+{
+    let mut target_cast = match typeconv::best_common_int_uint(ecx, types) {
+        Some(t) => t,
+        None => return,
+    };
+
+    // Speculatively rewrite all ints and uints to be of the target type.
+    let mut target_types = types.to_vec();
+    for t in target_types.iter_mut() {
+        match t {
+            Some(t) if t.is_signed_int() || t.is_unsigned_int() => *t = target_cast.clone(),
+            _ => {}
+        }
+    }
+
+    // Iff there is exactly one implementation that exactly matches the
+    // rewritten parameters, proceed. In more complex cases, e.g. polymorphic
+    // functions, polymorphic type resolution will handle the conversion in its
+    // own way.
+    if impls
+        .iter()
+        .filter(|i| i.params.exact_match_nullable(&target_types))
+        .count()
+        != 1
+    {
+        return;
+    }
+
+    *types = target_types;
+
+    for (arg, typ) in args.iter_mut().zip_eq(types) {
+        match (arg, typ) {
+            // Note that it is crucial here to use `if v == target_cast` because
+            // we want to only touch parameters whose types we've changed, or
+            // parameters whose types were already this type. If we used, e.g.
+            // `ScalarType::base_eq`, we could introduce inadvertent changes to
+            // the input type, e.g. rescaling numeric values.
+            (CoercibleScalarExpr::Coerced(expr), Some(v)) if v == &mut target_cast => {
+                let cast =
+                    typeconv::plan_cast(ecx, CastContext::Implicit, expr.clone(), &target_cast)
+                        .expect("casts known to be valid");
+                *expr = cast;
+            }
+            _ => {}
+        }
+    }
+}
```

Comment:
Would it be possible to instead extend the "best match logic" ([postgres docs](https://www.postgresql.org/docs/current/typeconv-func.html)) to have a Materialize extension step "4.c.i" defined as something like "Run through all candidates and keep those with the most 'near' matches on input types. Keep all candidates if none have near matches. If only one candidate remains, use it; else continue to the next step." And then we could define "near" matches as:
  * parameter type i32, actual type u16
  * parameter type i64, actual type u32
  * parameter type numeric, actual type u64

I think that would result in selecting the same function, but without needing to rewrite the types ahead of time?

Reply:
Let me type that up!

Reply:
Yeah, that works! Let me noodle on how best to express this relationship in a way that connects these two uses.